src/tunnel.cc

```diff
@@ -1128,10 +1128,18 @@ TunnelStateData::usePinned()
 {
     const auto serverConn = borrowPinnedConnection(request.getRaw());
     debugs(26,7, "pinned peer connection: " << serverConn);
-    if (!Comm::IsConnOpen(serverConn)) {
+
+    const char *fail = nullptr;
+    if (!Comm::IsConnOpen(serverConn))
+        fail = "pinned path failure";
+
+    if (request->pinnedConnection()->pinning.peerAccessDenied)
```

Comment:
Why not place this new check inside borrowPinnedConnection()? Are there any contexts where the denied connection can be borrowed?

If we want to detail the error, then borrowPinnedConnection() can throw an exception when it cannot give out a usable connection. The caller will catch and sendError() with the exception text as the last parameter.

Please note that the current IsConnOpen() check for the borrowed pinned connection is already kind of wrong/misleading -- borrowPinnedConnection() check openness and returns nil for closed connections, but it also returns nil in other cases, so calling IsConnOpen() again does not detect just closed connections, it detects all those other cases as well. The above adjustment can fix this (minor) problem as well.

Reply:
> Why not place this new check inside borrowPinnedConnection()? Are there any contexts where the denied connection can be borrowed?

This is the correct.
Fixed at commit 808f5fb

> If we want to detail the error, then borrowPinnedConnection() can throw an exception when it cannot give out a usable connection. The caller will catch and sendError() with the exception text as the last parameter.

This is requires a fully restructure of validatePinnedConnection/borrowPinnedConnection methods. As is these methods I am seeing easier to just add a new ConnStateData::pinning::error member which will keep the error description.

It is not a huge change, just I don't know if we should do it here just for a debug message. The exact error can be guessed from the other debug messages,

Reply:
Please review commit 9ff3f8f.

This is make the borrowPinnedConnection (and validatePinnedConnection) to throw on error.

I had to add a new static ConnStateData::BorrowPinnedConnection to handle the case the related ConnStateDara is gone.

I added a new class exception the PinningException which also holds an integer code of error and the string error description. Initially I planned to use std::system_error but this is designed to hold OS errors, in some platforms the code maybe is not an integer.

The PinningException::errType currently can get two values, one has the meaning that one of the sides of pinned connection is closed (so we should not need to return an error page) and the other means that the pinned connection can not be used for the given request because one or more rules violated and in this case we may return an error page.

Reply:
> The PinningException::errType currently can get two values, one has the meaning that one of the sides of pinned connection is closed (so we should not need to return an error page)

FYI: I failed to find PR code that disables error page sending in this case. In my rewrite, the error page is always returned. Perhaps you meant that we do not have to but still return an error page.

Also, we should return an error page if to-server (pinned) connection is closed. As for the from-client connection closure, we may still want to generate an error page for logging and similar purposes, even if that error page does not get to the client.

Reply:
I am resolving this discussion after rewriting/simplifying exception throwing/catching code.